akka-docs-dev/rst/scala/code/docs/stream/GraphStageDocSpec.scala

```diff
@@ -83,4 +87,423 @@ class GraphStageDocSpec extends AkkaSpec {
+      new GraphStageLogic(shape) {
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            val a = grab(in)
+            if (p(a)) push(out, a)
+            else pull(in)
+          }
+        })
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = {
+            pull(in)
+          }
+        })
+      }
+  }
+  //#many-to-one
+
+  "Demonstrate a many to one element GraphStage" in {
+
+    // tests:
+    val evenFilter = Flow.fromGraph(new Filter[Int](_ % 2 == 0))
+
+    val result =
+      Source(Vector(1, 2, 3, 4, 5, 6))
+        .via(evenFilter)
+        .runFold(Seq.empty[Int])((elem, acc) => elem :+ acc)
+
+    Await.result(result, 3.seconds) should ===(Seq(2, 4, 6))
+  }
+
+  //#one-to-many
+  class Duplicator[A] extends GraphStage[FlowShape[A, A]] {
+
+    val in = Inlet[A]("Duplicator.in")
+    val out = Outlet[A]("Duplicator.out")
+
+    val shape = FlowShape.of(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) {
+        // Again: note that all mutable state
+        // MUST be inside the GraphStageLogic
+        var lastElem: Option[A] = None
+
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            val elem = grab(in)
+            lastElem = Some(elem)
+            push(out, elem)
+          }
+
+          override def onUpstreamFinish(): Unit = {
+            if (lastElem.isDefined) emit(out, lastElem.get)
+            complete(out)
+          }
+
+        })
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = {
+            if (lastElem.isDefined) {
+              push(out, lastElem.get)
+              lastElem = None
+            } else {
+              pull(in)
+            }
+          }
+        })
+      }
+  }
+  //#one-to-many
+
+  "Demonstrate a one to many element GraphStage" in {
+    // tests:
+    val duplicator = Flow.fromGraph(new Duplicator[Int])
+
+    val result =
+      Source(Vector(1, 2, 3))
+        .via(duplicator)
+        .runFold(Seq.empty[Int])((elem, acc) => elem :+ acc)
+
+    Await.result(result, 3.seconds) should ===(Seq(1, 1, 2, 2, 3, 3))
+  }
+
+  "Demonstrate a simpler one to many stage" in {
+    //#simpler-one-to-many
+    class Duplicator[A] extends GraphStage[FlowShape[A, A]] {
+
+      val in = Inlet[A]("Duplicator.in")
+      val out = Outlet[A]("Duplicator.out")
+
+      val shape = FlowShape.of(in, out)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+        new GraphStageLogic(shape) {
+
+          setHandler(in, new InHandler {
+            override def onPush(): Unit = {
+              val elem = grab(in)
+              emitMultiple(out, Iterable(elem, elem))
+            }
+          })
+          setHandler(out, new OutHandler {
+            override def onPull(): Unit = {
+              pull(in)
```

Comment:
maybe comment that this handler is being "suspended" while the emitMultiple is running, and only given control back to it once that emitting has been finished.